sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,579 @@
+
+            let address = match ctxt.request {
+                MethodRequest::GetAccount((address,)) => *address,
+                request => panic!("programmer error: unexpected request: {:?}", request),
+            };
+
+            diem_root_with_proof
+                .verify(latest_li, state_version, diem_root_address())
+                .map_err(Error::invalid_proof)?;
+            account_state_with_proof
+                .verify(latest_li, state_version, address)
+                .map_err(Error::invalid_proof)?;
+
+            // TODO(philiphayes): it seems wasteful to lookup the whole diem_root
+            // account, verify the proofs, etc... just to get a list of supported
+            // currency codes. Would it make sense for the AccountView to just
+            // list _all_ BalanceResource's under balance? Then we could avoid
+            // this extra lookup. It's even more problematic when a batch
+            // includes many GetAccount requests, as this batch handling logic
+            // isn't (currently) smart enough to dedup the diem_root lookups.
+            let diem_root_blob = diem_root_with_proof
+                .blob
+                .ok_or_else(|| Error::unknown("missing diem_root account"))?;
+            let diem_root = AccountState::try_from(&diem_root_blob)
+                .map_err(Error::decode)?;
+            let currencies: Option<RegisteredCurrencies> = diem_root.get_config()
+                .map_err(Error::decode)?;
+            let currency_codes = currencies
+                .as_ref()
+                .map(RegisteredCurrencies::currency_codes)
+                .ok_or_else(|| Error::unknown("diem_root has no registered currencies"))?;
+
+            let maybe_account_view = account_state_with_proof
+                .blob
+                .map(|blob| {
+                    let account_state = AccountState::try_from(&blob)
+                        .map_err(Error::decode)?;
+                    AccountView::try_from_account_state(
+                        address,
+                        account_state,
+                        &currency_codes,
+                        state_version,
+                    ).map_err(Error::decode)
+                })
+                .transpose()?;
+
+            Ok(MethodResponse::GetAccount(maybe_account_view))
+        }
+        subresponses => return Err(Error::rpc_response(format!(
+                    "expected [GetAccountStateWithProof, GetAccountStateWithProof] subresponses, received: {:?}",
+                    subresponses,
+                    ))),
+    }
+    };
+    VerifyingRequest::new(request, subrequests, callback)
+}
+
+mod private {
+    pub trait Sealed {}
+
+    impl Sealed for super::InMemoryStorage {}
+}
+
+// TODO(philiphayes): unseal `Storage` trait once verifying client stabilizes.
+pub trait Storage: private::Sealed + Debug {
+    fn get(&self, key: &str) -> Result<Vec<u8>>;
+    fn set(&mut self, key: &str, value: Vec<u8>) -> Result<()>;
+}
```

Comment:
Is there a reason that you need to have the API like this where you supply a key when you're only using a single Key?

Reply:
To be honest, I'm just not sure what we need yet--the trait is mostly a place holder for now. I'm mostly imagining the on-disk storage as a directory with each key as a file name and the value as the byte contents. If I had to guess, I think we might get away with just storing the `TrustedState`, but maybe we also need to store some configs? chain id? frozen subtrees in the tx accumulator (or is this in the TrustedState)? Just leaving our options open for now.

Reply:
Fair enough, before we end up "stabilizing" this we'll need to figure out the correct abstraction then.